druid-shell/src/backend/wayland/display.rs

```diff
@@ -0,0 +1,215 @@
+    ) {
+        self.sub.consume(event, registry, ctx)
+    }
+}
+
+impl<X> From<X> for GlobalEventSubscription
+where
+    X: Fn(&wlc::GlobalEvent, &wlc::Attached<wl_registry::WlRegistry>, &wlc::DispatchData)
+        + GlobalEventConsumer
+        + 'static,
+{
+    fn from(closure: X) -> Self {
+        Self {
+            id: 0,
+            sub: std::sync::Arc::new(closure),
+        }
+    }
+}
+
+impl<X> GlobalEventConsumer for X
+where
+    X: Fn(&wlc::GlobalEvent, &wlc::Attached<wl_registry::WlRegistry>, &wlc::DispatchData) + 'static,
+{
+    fn consume(
+        &self,
+        event: &wlc::GlobalEvent,
+        registry: &wlc::Attached<wl_registry::WlRegistry>,
+        ctx: &wlc::DispatchData,
+    ) {
+        self(event, registry, ctx)
+    }
+}
+
+pub trait GlobalEventDispatch {
+    fn subscribe(&self, sub: impl Into<GlobalEventSubscription>) -> GlobalEventSubscription;
+    fn release(&self, s: &GlobalEventSubscription);
+}
+
+pub trait GlobalEventConsumer {
+    fn consume(
+        &self,
+        event: &wlc::GlobalEvent,
+        registry: &wlc::Attached<wl_registry::WlRegistry>,
+        ctx: &wlc::DispatchData,
+    );
+}
+
+pub(super) struct Dispatcher {
+    incr: crate::Counter,
+    subscriptions: std::cell::RefCell<BTreeMap<u64, GlobalEventSubscription>>,
+}
+
+impl Default for Dispatcher {
+    fn default() -> Self {
+        Self {
+            incr: crate::Counter::new(),
+            subscriptions: std::cell::RefCell::new(BTreeMap::new()),
+        }
+    }
+}
+
+impl GlobalEventConsumer for Dispatcher {
+    fn consume(
+        &self,
+        event: &wlc::GlobalEvent,
+        registry: &wlc::Attached<wl_registry::WlRegistry>,
+        ctx: &wlc::DispatchData,
+    ) {
+        // tracing::info!("global event initiated {:?} {:?}", registry, event);
+        for (_, sub) in self.subscriptions.borrow().iter() {
+            sub.consume(event, registry, ctx);
+        }
+        // tracing::info!("global event completed {:?} {:?}", registry, event);
+    }
+}
+
+impl GlobalEventDispatch for Dispatcher {
+    fn subscribe(&self, sub: impl Into<GlobalEventSubscription>) -> GlobalEventSubscription {
+        let sub = sub.into().with_id(self.incr.next());
+        self.subscriptions.borrow_mut().insert(sub.id, sub.clone());
+        sub
+    }
+
+    fn release(&self, s: &GlobalEventSubscription) {
+        self.subscriptions.borrow_mut().remove(&s.id);
+    }
+}
+
+pub(super) struct Environment {
+    pub(super) display: wlc::Display,
+    pub(super) registry: wlc::GlobalManager,
+    pub(super) xdg_base: wlc::Main<xdg_wm_base::XdgWmBase>,
+    pub(super) queue: std::rc::Rc<std::cell::RefCell<wlc::EventQueue>>,
+    dispatcher: std::sync::Arc<Dispatcher>,
+}
+
+// because we have the global environment we need to mark these as safe/send.
+// strictly speaking we should probably guard the access to the various fields
+// behind a mutex, but in practice we are not actually accessing across threads.
```

Comment:
can you explain this in more detail:
- where is this Send and Sync is needed? 
- how can we be sure that we are not using it across threads?

Reply:
entirely unnecessary now. will remove.